#  Getting Started With Python

Q1: Choose your base between 3 and 15

In [6]:
base = 15

In [8]:
print("Ans:", base)

Ans: 15


Q2: Write the 50 numbers in base 15.

Ans: 0,1,2,3,4,5,6,7,8,9,A,B,C,D,E,10,11,12,13,14,15,16,17,18,19,1A,1B,1C,1D,1E,20,21,22,23,24,25,26,27,28,29,2A,2B,2C,2D,2E,30,31,32,33,34

Q3: (i) Addition of single digits in base 15

Ans: 9 + A = 14, 5 + B = 11, 6 + E = 15.

(ii) Addition of multiple digits in base 15.

Ans: 1456 + 2ACE = 4035, 3EBD + 14AB = 5479, 9ADE + A3BC = 14EAB

Q4: (i) Subtraction of single digits in base 15

Ans: B - 8 = 3, D - 6 = 7, A - 9 = 1

(ii) Subtraction of multiple digits in base 15.

Ans: A1B2 - 9CDE = 3C3, E3D4 - 5A2B = 88A8, C2B1 - 4567 = 7C49

Q5: Prepare a table of single digit multiplication in base 15.

Ans: Base 15 Multiplication table (Single Digit)

|   | 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | A | B | C | D | E |
|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|
| 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 |
| 1 | 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | A | B | C | D | E |
| 2 | 0 | 2 | 4 | 6 | 8 | A | C | E | 11 | 13 | 15 | 17 | 19 | 1B | 1D |
| 3 | 0 | 3 | 6 | 9 | C | F | 12 | 15 | 18 | 1B | 1E | 21 | 24 | 27 | 2A |
| 4 | 0 | 4 | 8 | C | 10 | 14 | 18 | 1C | 20 | 24 | 28 | 2C | 30 | 34 | 38 |
| 5 | 0 | 5 | A | F | 14 | 19 | 1E | 23 | 28 | 2D | 32 | 37 | 3C | 41 | 46 |
| 6 | 0 | 6 | C | 12 | 18 | 1E | 24 | 2A | 30 | 36 | 3C | 42 | 48 | 4E | 54 |
| 7 | 0 | 7 | E | 15 | 1C | 23 | 2A | 31 | 38 | 3F | 46 | 4D | 54 | 5B | 62 |
| 8 | 0 | 8 | 11 | 18 | 20 | 28 | 30 | 38 | 40 | 48 | 50 | 58 | 60 | 68 | 70 |
| 9 | 0 | 9 | 13 | 1B | 24 | 2D | 36 | 3F | 48 | 51 | 5A | 63 | 6C | 75 | 7E |
| A | 0 | A | 15 | 1E | 28 | 32 | 3C | 46 | 50 | 5A | 64 | 6E | 78 | 82 | 8C |
| B | 0 | B | 17 | 21 | 2C | 37 | 42 | 4D | 58 | 63 | 6E | 79 | 84 | 8F | 9A |
| C | 0 | C | 19 | 24 | 30 | 3C | 48 | 54 | 60 | 6C | 78 | 84 | 90 | 9C | A8 |
| D | 0 | D | 1B | 27 | 34 | 41 | 4E | 5B | 68 | 75 | 82 | 8F | 9C | A9 | B6 |
| E | 0 | E | 1D | 2A | 38 | 46 | 54 | 62 | 70 | 7E | 8C | 9A | A8 | B6 | C4 |

Q6: Multiply multiple digits in base 15.

Ans: A1B2 * 9CDE = 69B70D1D, E3D4 * 5A2B = 55E9243E, C2B1 * 4567 = 3821B087

Q7: Convert the following numbers from base 15 to base 10.

 (i) 1 
 
 Ans: 1
 
 (ii) 10

 Ans: 15
 
 (iii) 20
 
 Ans: 30
 
 (iv) 30
 
 Ans: 45
 
 (v) 100
 
 Ans: 225

Q8: Write a strategy to convert a number from base 15 to base 10 in
 (i) plain english
 (ii) pseudo code
 (iii) code

Ans: (i) plain english

 1. In base 15, we have the following digits: 0,1,2,3,4,5,6,7,8,9,A,B,C,D,E,
 
 which in base 10, the digits
 
 0-9 are represented by 0-9 themselves
 
 A is represented by 10
 
 B is represented by 11
 
 C is represented by 12
 
 D is represented by 13
 
 E is represented by 14
 
 
 2. Now, the positional values of the digits vary as powers of 15, i.e., 
 
 (a) For the rightmost digit, the positional value is 15^0.
 
 (b) For the next digit left to the rightmost digit, the positional value is 15^1.
 
 (c) For the next digit to the left of the digit with positional value 15^1, the positional value is 15^2, ans so on....
 
 
 3. Now, we convert each digit in the number of base 15 to the decimal equivalent of the respective digits.
 
 
 4. We multiply each converted decimal equivalent of the number of base 15 by its corresponding positional value.
 
 
 5. We add all the resulting products together to obtain the desired number(initially in base 15) in base 10.
 
 
 (ii) pseudo code
 
 
    define function find_index(array, element):
   
            #input: 'array' represents the array of elements, 'element' represents the element to be found
   
            #output: The index of the element in the array, or 'NOT FOUND' if not found
            
            For i from 0 to len(array)-1:
                   
                   If array[i] == element
                         
                         return i 
                         
                    else: 
                    
                          return None
                          
      define function base15_to_base10(number):
       
              #input: base 15 number
              
              #output: base 10 number 
              
              syms_of_base15 = "0123456789ABCDE"
              
              base10valueofinput = 0
              
              power=0
              
              For i from len[str(number)]-1 to 0:
                      
                      digit = str(number)[i]
                      
                      base10digitvalue = find_index(syms_of_base15, digit)
                      
                        if base10digitvalue == None
                         
                         return "invalid input"
                         
                        else:
                          
                          base10valueofinput = base10valueofinput + base10digitvalue* (15^power)
                          
                          power = power +1
                          
                         return base10valueofinput
                         
                         
(iii) code

In [3]:
def find_index(array, element):
    for i in range(len(array)):
        if array[i] == element:
            return i
    return None

def base15_to_base10(number):
    syms_of_base15 = "0123456789ABCDE"
    base10valueofinput = 0
    power = 0

    number_str = str(number)
    for i in range(len(number_str) - 1, -1, -1):
        digit = number_str[i]
        base10digitvalue = find_index(syms_of_base15, digit)

        if base10digitvalue is None:
            return "invalid input"
        else:
            base10valueofinput = base10valueofinput + base10digitvalue * (15 ** power)
            power += 1
    return base10valueofinput

In [5]:
result = base15_to_base10(5287)
print("Final result:", result)

Final result: 17452


In [3]:
result = base15_to_base10("5AB78C4")
print("Final result:", result)

Final result: 65129359


In [6]:
result = base15_to_base10("fghg")
print("Final result:", result)

Final result: invalid input


Q9: Write a strategy to convert a number in base N to base M in (i) plain english  (ii) pseudo code  (iii) code

Q10: In base 10, how many numbers can you represent in 5 digits?

In [1]:
10**5

100000

Q11: In base 7, how many numbers can you represent in 5 digits?

In [2]:
7**5

16807

Q12: In base B, how many numbers can you represent in n digits?

Ans: B^n

Q13: To represent numbers upto 10000 but not including 10000 in base 10, how many digits do you need ?

Ans: 4


Note: We are representing total 10000 numbers from 0 to 9999.

Q14: To represent N numbers, how many digits do we need in base 10?

Ans: d = log N

Q15: To represent N numbers, how many digits do we need in base B?

Ans: d = log(B) N

# Logarithm: Finding log base 10 of any number 

In [1]:
import math

In [54]:
math.log(528) #base is e

6.269096283706261

In [3]:
number = 528

We know 10^2 < 528 < 10^3

Q16: Write a pseudo code or an algorithm to calculate log(10) 528.

Ans: We know log(10)528 lies between 2(left) and 3(right). Thus, we take the average of the two values (that is, the average of 2 and 3) as the 'new value' and calculate 10^(new value).
   
   If 10^(new value) < 528, we take our 'left' = 'new value' and keep 'right' as it is and then take the average again of 'new value' ans 'right' and so on....
   
   But if 10^(new value) > 528, we take our 'new value' = 'right' and keep 'left' as it is and repeat the same process of averaging and proceeding further...
   
   We repeat the above steps till we get a precise answer.



In [4]:
10**2.5 # 2.5 is the average of 2 and 3

316.22776601683796

We see 316.22776601683796   < 528, therefore left = 2.5 and right = 3.
Again, average of 2.5 and 3 is 2.75. Thus, we calcualte 10^(2.75).


In [10]:
10**2.75

562.341325190349

Again, our result 562.341325190349 > 528, therefore our new right = 2.75 and left = 2.5. And thus, we calculate 10^[(2.5 + 2.75)/2] and continue the process till we get an accurate answer.

In [11]:
left = 2
right = 3
mid = (left+right)/2
10**mid

316.22776601683796

In [12]:
mid

2.5

In [13]:
left = mid
left, right

(2.5, 3)

In [14]:
print(f"Left: {left}")
print(f"Right: {right}")
mid = (left+right)/2
print(f"Mid: {mid}")
print(10**mid)

Left: 2.5
Right: 3
Mid: 2.75
562.341325190349


In [15]:
right = mid

In [16]:
print(f"Left: {left}")
print(f"Right: {right}")
mid = (left+right)/2
print(f"Mid: {mid}")
print(10**mid)

Left: 2.5
Right: 2.75
Mid: 2.625
421.6965034285823


In [17]:
left = mid

In [18]:
print(f"Left: {left}")
print(f"Right: {right}")
mid = (left+right)/2
print(f"Mid: {mid}")
print(10**mid)

Left: 2.625
Right: 2.75
Mid: 2.6875
486.9675251658631


In [19]:
left = mid

In [20]:
print(f"Left: {left}")
print(f"Right: {right}")
mid = (left+right)/2
print(f"Mid: {mid}")
print(10**mid)

Left: 2.6875
Right: 2.75
Mid: 2.71875
523.2991146814946


In [21]:
left = mid

In [22]:
print(f"Left: {left}")
print(f"Right: {right}")
mid = (left+right)/2
print(f"Mid: {mid}")
print(10**mid)

Left: 2.71875
Right: 2.75
Mid: 2.734375
542.4690937011326


In [23]:
right = mid

In [24]:
print(f"Left: {left}")
print(f"Right: {right}")
mid = (left+right)/2
print(f"Mid: {mid}")
print(10**mid)

Left: 2.71875
Right: 2.734375
Mid: 2.7265625
532.7978945865641


In [25]:
right = mid

In [26]:
print(f"Left: {left}")
print(f"Right: {right}")
mid = (left+right)/2
print(f"Mid: {mid}")
print(10**mid)

Left: 2.71875
Right: 2.7265625
Mid: 2.72265625
528.0271456481318


Thus, we can see, 528.0271456481318 is approximately equal to 528.

In [27]:
left = 2
right = 3
number = 528

In [53]:
print(f"Left: {left}")
print(f"Right: {right}")
mid = (left+right)/2
print(f"Mid: {mid}")
result = 10**mid
print(f"Result: {result}")
diff = abs(result - number)
print(f"Diff: {diff}")
if diff <= 0.0001:
    print(f"Success!! log(10){number}: {mid}", )
else:
    if result > number:
        right = mid
    else:
        left = mid

Left: 2.7226338386535645
Right: 2.7226340770721436
Mid: 2.722633957862854
Result: 528.0000429518117
Diff: 4.295181167890405e-05
Success!! log(10)528: 2.722633957862854


In [55]:
left = 2
right = 3
number = 528

In [78]:
print(f"Left: {left}")
print(f"Right: {right}")
mid = (left+right)/2
print(f"Mid: {mid}")
result = 10**mid
print(f"Result: {result}")
diff = abs(result - number)
print(f"Diff: {diff}")
if diff <= 0.0001:
    print(f"Success!! log(10){number}: {mid}", )
else:
    if result > number:
        right = mid
    else:
        left = mid

Left: 2.7226338386535645
Right: 2.7226340770721436
Mid: 2.722633957862854
Result: 528.0000429518117
Diff: 4.295181167890405e-05
Success!! log(10)528: 2.722633957862854


In [79]:
left, right = 2, 3
number = 528

In [80]:
while True:
    print(f"Left: {left}")
    print(f"Right: {right}")
    mid = (left+right)/2
    print(f"Mid: {mid}")
    result = 10**mid
    print(f"Result: {result}")
    diff = abs(result - number)
    print(f"Diff: {diff}")
    if diff <= 0.0001:
        print(f"Success!! log10{number}: {mid}", )
        break
    else:
        if result > number:
            right = mid
        else:
            left = mid

Left: 2
Right: 3
Mid: 2.5
Result: 316.22776601683796
Diff: 211.77223398316204
Left: 2.5
Right: 3
Mid: 2.75
Result: 562.341325190349
Diff: 34.34132519034904
Left: 2.5
Right: 2.75
Mid: 2.625
Result: 421.6965034285823
Diff: 106.30349657141772
Left: 2.625
Right: 2.75
Mid: 2.6875
Result: 486.9675251658631
Diff: 41.032474834136906
Left: 2.6875
Right: 2.75
Mid: 2.71875
Result: 523.2991146814946
Diff: 4.700885318505357
Left: 2.71875
Right: 2.75
Mid: 2.734375
Result: 542.4690937011326
Diff: 14.469093701132579
Left: 2.71875
Right: 2.734375
Mid: 2.7265625
Result: 532.7978945865641
Diff: 4.797894586564098
Left: 2.71875
Right: 2.7265625
Mid: 2.72265625
Result: 528.0271456481318
Diff: 0.027145648131750022
Left: 2.71875
Right: 2.72265625
Mid: 2.720703125
Result: 525.657814405402
Diff: 2.3421855945980496
Left: 2.720703125
Right: 2.72265625
Mid: 2.7216796875
Result: 526.841148096957
Diff: 1.1588519030430007
Left: 2.7216796875
Right: 2.72265625
Mid: 2.72216796875
Result: 527.4338135156115
Diff: 0.566186

Q17: Write a code to calculate log(10)528 using function.

In [88]:
def log10(number):
    left = 0
    right = 10
    while True:
        mid = (left+right)/2
        result = 10**mid
        diff = abs(result - number)
        if diff <= 0.0001:
            return mid
        else:
            if result > number:
                right = mid
            else:
                left = mid

In [89]:
log10(528)

2.7226339280605316

In [90]:
log10(10)

1.0000038146972656

In [83]:
math.log10(528)

2.722633922533812

The above method of calculating log(10) number is called binary search.